Add File tutorial. (#2)
realm/unity-examples: tutorials/persistance-comparison-series/File/file.md

# Saving Data in Unity3D using Files
<b>(Part 2 of the Persistence Comparison Series)</b>

## Introduction

Persisting data is an important part of most games. Unity offers only a limited set of solutions which means we have to look around for other options as well.

In Part 1 of this series we explored Unity's own solution: `PlayerPrefs`. This time we look into one of the ways we can use the underlying .NET framework by saving files. Here is an overview of the complete series:

- Part 1: [PlayerPrefs](https://github.com/realm/unity-examples/blob/persistence-comparison/tutorials/persistance-comparison-series/PlayerPrefs/player_prefs.md)
- Part 2: Files *(this tutorial)*
- Part 3: BinaryReader and BinaryWriter *(coming soon)*
- Part 4: SQL
- Part 5: Realm Unity SDK
- Part 6: Comparison of all those options

> Like part 1 this tutorial can also be found in the https://github.com/realm/unity-examples repository on the [persistence-comparison](https://github.com/realm/unity-examples/tree/persistence-comparison) branch.

Reviewer: Should the branch be https://github.com/realm/unity-examples/tree/df/persistence-comparison-file or the plan is to merge everything in https://github.com/realm/unity-examples/tree/df/persistence-comparison?

Author: The plan is to merge them all into `persistent-comparison` so that the whole series lives in one branch and is easy accessible. Also, since the same game is used, it's kinda nice to see them all work side by side and compare them.

> Provides static methods for the creation, copying, deletion, moving, and opening of a single file, and aids in the creation of FileStream objects.

Besides that the `File` class is also used to manipulate the file itself, reading and writing data. On top of that if offers ways to read meta data of a file like time of creation.

When working with a file you can also make use of several options to change `FileMode` or `FileAccess.`

The `FileStream` mentioned in the documentation is another approach to work with those files, providing additional options. In this tutorial we will just use the plain `File` class.

Let's have a look at what we have to change in the example presented in the previous section to save the data using `File`:

```cs
using System;
using System.IO;
using UnityEngine;

public class FileExampleSimple : MonoBehaviour
{
    // Resources:
    // https://docs.microsoft.com/en-us/dotnet/api/system.io.file?view=net-5.0

    [SerializeField] private int hitCount = 0;

    private const string HitCountFile = "hitCountFile.txt"; // 1

    private void Start() // 5
    {
        if (File.Exists(HitCountFile)) // 6
        {

>             string textFileWriteAllText = File.ReadAllText(HitCountFile); // 7

Reviewer: This is a strange variable name

Author: 61e973c

Whenever we click on the capsule (2) and increment the hit count (3) we need to save that change. Using `File.WriteAllText()` (4) the file will be opened, data will be saved and it will be closed right away. Besides the file name this function expects the contents as a string. Therefore we have to transform the `hitCount` by calling `ToString()` before passing it on.

The next time we start the game (5) we want to load the previously saved data. First we check if the file already exists (6). If it does not exist we did never save before and can just keep the default value for `hitCount`. If the file exists we use `ReadAllText()` to get that data (7). Since this is a string again, we need to convert here as well using `Int32.Parse()` (8). Note that this means we have to be sure about what we read. If the structure of the file changes or the player edits it this might lead to problems during the parsing of the file.

Let's look into extending this simple example in the next section.

## Extended example

(see `FileExampleExtended.cs` in the repository for the finished version)

> The previous section has shown the most simple example, using just one variable that needs to be saved. What if we want to save more than that?

Reviewer: ```suggestion
The previous section has shown the simplest example, using just one variable that needs to be saved. What if we want to save more than that?
```

    {
        // Set the LeftShift key.
        keyPressed = KeyCode.LeftShift; // 3
    }
    else if (Input.GetKey(KeyCode.LeftControl)) // 2
    {
        // Set the LeftControl key.
        keyPressed = KeyCode.LeftControl; // 3
    }
    else // 4
    {
        // In any other case reset to default and consider it unmodified.
        keyPressed = default; // 5
    }
}
```

Now to saving the data when a click happens:

```cs
private void OnMouseDown() // 6
{
    // Check if a key was pressed.
    if (keyPressed == KeyCode.LeftShift) // 7
    {
        // Increment the Shift hit count.
        hitCountShift++; // 8
    }
    else if (keyPressed == KeyCode.LeftControl) // 7
    {
        // Increment the Control hit count.
        hitCountControl++; // 8
    }
    else // 9
    {
        // If neither Shift nor Control was held, we increment the unmodified hit count.
        hitCountUnmodified++; // 10
    }

    // 11
    // Create a string array with the three hit counts.
    string[] stringArray = {
        hitCountUnmodified.ToString(),
        hitCountShift.ToString(),
        hitCountControl.ToString()
    };

    // 12
    // Save the entries, line by line.

>     File.WriteAllLines(HitCountFileUnmodified, stringArray);

Reviewer: I don't think we're showing an example of how to read the extended file format.

Author: Indeed. I turned the structure around here since I thought it might be nicer to not follow the game flow but first save and then read (we only ever start the game after everything is done anyway) and then I must have forgotten to read it after saving it. :D

9a5844e

Check if `keyPressed` was set to `KeyCode.LeftShift` or `KeyCode.LeftControl` (7) and if so, increment the corresponding hit count (8). If no modifier was used (9), increment the `hitCountUnmodified`.

As seen in the last section we need to create a string that can be saved in the file. There is a second function on `File` that accepts a a string array and then saves each entry in one line: `WriteAllLines()`.

Knowing this we create an array containing the three hit counts (11) and pass this one on to `File.WriteAllLines()`.

Start the game, click the capsule using Shift and Control. You should see the three counters in the Inspector.

<img src="images/07_hit_count_extended_editor.jpg" alt="File hitCountExtended.txt" width=""/>

After stopping the game and therefore saving the data a new file `hitCountFileExtended.txt` should exist in your project folder. Have look at it, it should look something like this:

<img src="images/08_hit_count_extended_file.jpg" alt="File hitCountExtended.txt" width=""/>

> As long as all the data saved to those lines belongs together, the file can be one option. If you have several different properties you might create multiple files. Alternatively you can save all the data into the same file using a bit of structure. One of the possible approaches to structuring your data will be shown in the next section.

Reviewer: We should mention that the downside of this approach is that the numbers are not associated with the actual properties, which means that if someone refactors the code and reorders the checks, they'd be reading incorrect values.

Author: a8c02d9

}
```

The class itself can be `private` and just be added inside the `FileExampleJson` class but its fields need to be public.

As before we use a different file to save this data. Update the filename to:

```cs
private const string HitCountFileJson = "hitCountFileJson.txt";
```

When saving the data we will use the same `Update()` method as before to detect which key was pressed.

The first part of `OnMouseDown()` (1) can stay the same as well since this part only increments the hit count in depending on the modifier used.

```cs
private void OnMouseDown()
{
    // 1
    // Check if a key was pressed.
    if (keyPressed == KeyCode.LeftShift)
    {
        // Increment the Shift hit count.
        hitCountShift++;
    }
    else if (keyPressed == KeyCode.LeftControl)
    {
        // Increment the Control hit count.
        hitCountControl++;
    }
    else
    {
        // If neither Shift nor Control was held, we increment the unmodified hit count.
        hitCountUnmodified++;
    }

    // 2
    // Create a new HitCount object to hold this data.
    HitCount hitCount = new();
    hitCount.Unmodified = hitCountUnmodified;
    hitCount.Shift = hitCountShift;

>     hitCount.Control = hitCountControl;

Reviewer: This can be simplified similar to the previous example

Author: done

    // 3
    // Create a JSON using the HitCount object.
    string jsonString = JsonUtility.ToJson(hitCount, true);

    // 4
    // Save the json to the file.

>     File.WriteAllText(HitCountFileJson, jsonString);

Reviewer: This section is also missing the code for reading the json data and initializing the values.

Author: Also included in 9a5844e

}
```

However, we need to update the second part. Instead of a string array we create a new `HitCount` object and set the three public fields to the values of the hit counters (2).

Using `JsonUtility.ToJson()` we can transform this object to a string (3). If you pass in `true` for the second, optional parameter `prettyPrint` the string will be formatted in a nicely readable way.

Finally, as in `FileExampleSimple.cs` we just use `WriteAllText()` since we're only saving one string, not an array.

When you run the game you will the that in the editor it looks identical to the previous section since we are using the same three counters. If you open the file `hitCountFileJson.txt` you should then see the three counters in a nicely formatted JSON.

<img src="images/09_hit_count_extended_json.jpg" alt="File hitCountExtended.txt" width=""/>

## Conclusion

In this tutorial we learned how to utilize `File` to save data. `JsonUtility` helps structure this data. They are simple and easy to use, not much code is required.

What are the down sides though?

> First of, we open, write to and save the file every single time the capsule it clicked. While not a problem in this case and certainly applicable for some games this will not perform very well when many save operations are made.

Reviewer: ```suggestion
First of all, we open, write to, and save the file every single time the capsule it clicked. While not a problem in this case and certainly applicable for some games this will not perform very well when many save operations are made.
```

> Also, the data is saved in plain text and can easily be edited by the player.

Reviewer: You can probably extend the json example by adding encryption as a way to solve this.

> The more complex your data is, the more complex it will be to actually maintain this approach. What if the structure of the `HitCount` object changes? You have to change account for that when loading an older version of the JSON. Migrations are necessary.

Reviewer: ```suggestion
The more complex your data is, the more complex it will be to actually maintain this approach. What if the structure of the `HitCount` object changes? You have to account for that when loading an older version of the JSON. Migrations are necessary.
```

    switch (modifier)
    {
        case KeyCode.LeftShift: // 7
            // Increment the Shift hit count.
            hitCountShift++; // 8
            break;
        case KeyCode.LeftCommand: // 7
            // Increment the Control hit count.
            hitCountControl++; // 8
            break;
        default: // 9
            // If neither Shift nor Control was held, we increment the unmodified hit count.
            hitCountUnmodified++; // 10
            break;
    }

    // 11
    // Create a string array with the three hit counts.
    string[] stringArray = {
        hitCountUnmodified.ToString(),
        hitCountShift.ToString(),
        hitCountControl.ToString()
    };

    // 12
    // Save the entries, line by line.
    File.WriteAllLines(HitCountFileUnmodified, stringArray);
}
```

Whenever a mouse click is detected on the capsule (6) we can then perform a similar check to what happened in `Update()` only that we use `modifier` instead of `Input.GetKey()` here.

Check if `modifier` was set to `KeyCode.LeftShift` or `KeyCode.LeftControl` (7) and if so, increment the corresponding hit count (8). If no modifier was used (9), increment the `hitCountUnmodified`.

> As seen in the last section we need to create a string that can be saved in the file. There is a second function on `File` that accepts a a string array and then saves each entry in one line: `WriteAllLines()`.

Reviewer 2: ```suggestion
As seen in the last section we need to create a string that can be saved in the file. There is a second function on `File` that accepts a string array and then saves each entry in one line: `WriteAllLines()`.
```
